myia/vm.py

```diff
@@ -0,0 +1,222 @@
+        """
+        frame = self
+        while frame and node.graph and node.graph is not frame.graph:
+            frame = frame.closure
+        if node not in frame.values:
+            # we set up a fake node to jump back to this frame after
+            # the value that we need has been evaluated.
+            if frame is not self:
+                frame.todo.append(self.Jump(self))
+            frame.todo.append(node)
+            if frame is not self:
+                raise self.Return(frame)
+            else:
+                raise self.Continue()
+        return frame.values[node]
+
+    def wrap_closure(self, value):
+        """Wrap graphs that are closures."""
+        if isinstance(value, Graph):
+            if any(n.graph and n.graph is not value
+                   for n in dfs(value.return_)):
+                return self.Closure(value, self)
+        return value
+
+    def do_call(self, graph: Union[Graph, 'VMFrame.Closure'], args: List[Any]):
+        """Perform a call.
+
+        This will handle tail calls and closures.
+
+        """
+        if self.tail() and self.parent is not None:
+            parent = self.parent
+        else:
+            parent = self
+        if isinstance(graph, self.Closure):
+            new_frame = VMFrame(graph.graph, args, parent, graph.frame)
+        else:
+            new_frame = VMFrame(graph, args, parent)
+        raise self.Return(new_frame)
+
+    def advance(self) -> 'VMFrame':
+        """
+        Take a step.
+
+        This will perform one "step" of execution.  This corresponds
+        roughly to evaluating one node of the graph.  In the case of
+        function calls, we single-step through the called function
+        even though the apply is a single node in our graph.
+        """
+        if self.done():
+            raise StopIteration()
+
+        while True:
+            node = self.todo[-1]
+            assert node not in self.values
+
+            try:
+                self.eval_node(node)
+            except self.Continue:
+                continue
+            except self.Return as e:
+                return e.value
+
+            self.todo.pop()
+            return None
+
+    def eval_node(self, node: ANFNode):
+        """
+        Compute the value for a node.
+
+        This may fail and ask for a missing value to be computed or
+        enter into a call.
+        """
+        if isinstance(node, Constant):
+            self.values[node] = self.wrap_closure(node.value)
+        elif isinstance(node, Parameter):
+            idx = self.graph.parameters.index(node)
+            self.values[node] = self.args[idx]
+        elif isinstance(node, self.Jump):
+            self.todo.pop()
+            raise self.Return(node.value)
+        elif isinstance(node, Apply):
+            fn = self.get_value(node.inputs[0])
+            args = [self.get_value(i) for i in node.inputs[1:]]
+            if isinstance(fn, Primitive):
```

Comment:
We need to figure out where to put primitive implementations. I'd suggest a dictionary mapping primitives to functions, so that we can easily experiment with different implementations. One slight complication is that the implementation for `if`, as it stands, requires the VM.

Reply:
We can special case If and have a dict or something for the rest.  I've implemented add inline for now, but I'll move it when we have more.